swarm/storage/types.go

```diff
@@ -285,3 +311,74 @@ func (c ChunkData) Size() int64 {
 func (c ChunkData) Data() []byte {
 	return c[8:]
 }
+
+func NoValidateChunk(hasher SwarmHash, key *Key, data []byte) bool {
+	return true
+}
+
+// Provides method for validation of content address in chunks
+// Holds the corresponding hasher to create the address
+type ContentAddressValidator struct {
+	hashMu sync.Mutex
+	Hasher SwarmHash
+}
+
+// Constructor
+func NewContentAddressValidator(hash SwarmHash) *ContentAddressValidator {
+	return &ContentAddressValidator{
+		Hasher: hash,
+	}
+}
+
+// Validate that the given key is a valid content address for the given data
+func (self *ContentAddressValidator) Validate(key Key, data []byte) bool {
+	self.hashMu.Lock()
+	defer self.hashMu.Unlock()
+	self.Hasher.Reset()
+	self.Hasher.Write(data)
+	hash := self.Hasher.Sum(nil)
+
+	if !bytes.Equal(hash, key[:]) {
+		log.Error("invalid content address. Expected %x, have %x", hash, key)
+		return false
+	}
+	return true
+}
+
+// Common signature for chunk validation functions
+type ChunkValidatorFunc func(Key, []byte) bool
+
+// Provides method for validating any chunk type
+type ChunkValidator struct {
+	content  ChunkValidatorFunc
+	resource ChunkValidatorFunc
+}
+
+// Constructor
+func NewChunkValidator(content ChunkValidatorFunc, resource ChunkValidatorFunc) *ChunkValidator {
+	return &ChunkValidator{
+		content:  content,
+		resource: resource,
+	}
+}
+
+// Validate the integrity of the chunk
+//
+// First checks if the key is a valid content address, as in SwarmHash(data)
+// If this fails, it checks if the chunk data is a valid resource update chunk
+//
+// If resource is nil and content is not, then check will fail if content check fails
+// If content is nil and resource is not, then check falls through to resource
+// If both are nil, all chunks will be valid
+func (self *ChunkValidator) Validate(key Key, data []byte) bool {
+	if self.content != nil {
```

Comment:
I think that a design where we have `ValidateContent` and `ValidateResource` would be better. Then you are free to implement this interface as you see fit, and no need for `nil` checks that change behaviour.

Reply:
Having discussed this a few times, I think we should replace this with a simpler interface, like:

https://github.com/nonsense/go-ethereum/commit/1b597764beea555da709a703b3ba954ac56cf247